daemon/api_model.go

```diff
@@ -50,51 +57,147 @@ var (
 	}
 )
 
-var devicestateRemodel = devicestate.Remodel
+var (
+	devicestateRemodel = devicestate.Remodel
+	sideloadSnapsInfo  = sideloadInfo
+)
 
 type postModelData struct {
 	NewModel string `json:"new-model"`
 }
 
 func postModel(c *Command, r *http.Request, _ *auth.UserState) Response {
 	contentType := r.Header.Get("Content-Type")
-	mediaType, _, err := mime.ParseMediaType(contentType)
+	mediaType, params, err := mime.ParseMediaType(contentType)
 	if err != nil {
 		// assume json body, as type was not enforced in the past
 		mediaType = "application/json"
 	}
 
 	switch mediaType {
 	case "application/json":
+		// If json content type we get only the new model assertion and
+		// the rest is downloaded from the store.
 		return storeRemodel(c, r)
 	case "multipart/form-data":
-		return BadRequest("media type %q not implemented yet", mediaType)
+		// multipart/form-data content type can be used to sideload
+		// part of the things necessary for a remodel.
+		return offlineRemodel(c, r, params)
 	default:
 		return BadRequest("unexpected media type %q", mediaType)
 	}
 }
 
+func modelFromData(data []byte) (*asserts.Model, error) {
+	rawNewModel, err := asserts.Decode(data)
+	if err != nil {
+		return nil, fmt.Errorf("cannot decode new model assertion: %v", err)
+	}
+	newModel, ok := rawNewModel.(*asserts.Model)
+	if !ok {
+		return nil, fmt.Errorf("new model is not a model assertion: %v", newModel.Type())
+	}
+
+	return newModel, nil
+}
+
 func storeRemodel(c *Command, r *http.Request) Response {
 	var data postModelData
 	decoder := json.NewDecoder(r.Body)
 	if err := decoder.Decode(&data); err != nil {
 		return BadRequest("cannot decode request body into remodel operation: %v", err)
 	}
+	newModel, err := modelFromData([]byte(data.NewModel))
+	if err != nil {
+		return BadRequest(err.Error())
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
 
-	rawNewModel, err := asserts.Decode([]byte(data.NewModel))
+	chg, err := devicestateRemodel(st, newModel, nil, nil)
 	if err != nil {
-		return BadRequest("cannot decode new model assertion: %v", err)
+		return BadRequest("cannot remodel device: %v", err)
 	}
-	newModel, ok := rawNewModel.(*asserts.Model)
-	if !ok {
-		return BadRequest("new model is not a model assertion: %v", newModel.Type())
+	ensureStateSoon(st)
+
+	return AsyncResponse(nil, chg.ID())
+}
+
+func offlineRemodel(c *Command, r *http.Request, contentTypeParams map[string]string) Response {
```

Comment:
Personal preference: From a code readability/maintainability perspective this function is doing way to many things. It's responsible for reading the form data, extracting data it needs, committing data to a DB, build snap sideinfo data and then schedule the remodel tasks. The function even have subheaders to mark different sections in the function.